File: typescript/Day3.ts
import { BaseDay } from "./BaseDay";

class Day3 extends BaseDay {
    part1(puzzleInput: string[]): string {
        let length = puzzleInput[0].length;
        let zeros: Array<number> = [];

        let gammaRate = "";
        let epsilonRate = "";

        for (let i = 0; i < puzzleInput.length; i++) {
            for (let j = 0; j < length; j++) {
                let number = puzzleInput[i][j];
                if (number == '0') {
                    if (zeros[j]) {
                        zeros[j]++;
                    } else {
                        zeros[j] = 1;
                    }
                }
            }
        }

        for (let i = 0; i < length; i++) {
            if (zeros[i] > Math.floor(puzzleInput.length / 2)) {
                gammaRate += "0";
                epsilonRate += "1";
            } else {
                gammaRate += "1";
                epsilonRate += "0";
            }
        }

        return (parseInt(gammaRate, 2) * parseInt(epsilonRate, 2)).toString();
    }

    part2(puzzleInput: string[]): string {
        let oxygenGeneratorRating = this.getOxygenGeneratorRating(puzzleInput);
        let co2ScrubberRating = this.getCo2ScrubberRating(puzzleInput);
        return (oxygenGeneratorRating * co2ScrubberRating).toString();
    }

    getCo2ScrubberRating(puzzleInput: Array<string>) {
        let list: Array<string> = this.getListOfLessOccurs(puzzleInput);
        return parseInt(list[0], 2);
    }

    getOxygenGeneratorRating(puzzleInput: Array<string>) {
        let list: Array<string> = this.getListOfMostOccurs(puzzleInput);
        return parseInt(list[0], 2);
    }

    getListOfMostOccurs(list: Array<string>, position: number = 0): Array<string> {
        return this.getListOf(list, position, (countZeros: number, countOnes: number) => countZeros > countOnes);
    }

    getListOfLessOccurs(list: Array<string>, position: number = 0): Array<string> {
        return this.getListOf(list, position, (countZeros: number, countOnes: number) => countZeros <= countOnes);
    }

    getListOf(list: Array<string>, position: number, listFilter: ListFilter): Array<string> {
        let numbersWithZeros: Array<string> = [];
        let numbersWithOnes: Array<string> = [];

        for (let i = 0; i < list.length; i++) {
            let number = list[i][position];
            if (number == '0') {
                numbersWithZeros.push(list[i]);
            } else {
                numbersWithOnes.push(list[i]);
            }
        }

        let countZeros = numbersWithZeros.length;
        let countOnes = numbersWithOnes.length;
        let nextList: Array<string> = listFilter(countZeros, countOnes) ? numbersWithZeros : numbersWithOnes;

        if (nextList.length == 1) {
            return nextList;
        } else {
            return this.getListOf(nextList, position + 1, listFilter);
        }
    }
}

interface ListFilter {
    (countZeros: number, countOnes: number): boolean;
}

new Day3().run(3);

File: java/Day3.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Day3 extends Day {
    @Override
    int part1(String[] puzzleInput) {
        int charLength = puzzleInput[0].length();
        int[] zeros = new int[charLength];

        StringBuilder gammaRate = new StringBuilder();
        StringBuilder epsilonRate = new StringBuilder();

        for (int i = 0; i < puzzleInput.length; i++) {
            for (int j = 0; j < charLength; j++) {
                char number = puzzleInput[i].charAt(j);
                if (number == '0') {
                    zeros[j]++;
                }
            }
        }

        for (int i = 0; i < charLength; i++) {
            if (zeros[i] > Math.floor(puzzleInput.length / 2)) {
                gammaRate.append("0");
                epsilonRate.append("1");
            } else {
                gammaRate.append("1");
                epsilonRate.append("0");
            }
        }
        
        return Integer.parseInt(gammaRate.toString(), 2) * Integer.parseInt(epsilonRate.toString(), 2);
    }

    @Override
    int part2(String[] puzzleInput) {
        int oxygenGeneratorRating = getOxygenGeneratorRating(puzzleInput);
        int co2ScrubberRating = getCo2ScrubberRating(puzzleInput);
        return oxygenGeneratorRating * co2ScrubberRating;
    }

    private int getCo2ScrubberRating(String[] puzzleInput) {
        List<String> list = getListOfLessOccurs(puzzleInput);
        return Integer.parseInt(list.get(0), 2);
    }

    private int getOxygenGeneratorRating(String[] puzzleInput) {
        List<String> list = getListOfMostOccurs(puzzleInput);
        return Integer.parseInt(list.get(0), 2);
    }

    private List<String> getListOfMostOccurs(String[] list) {
        return getListOfMostOccurs(Arrays.asList(list), 0);
    }

    private List<String> getListOfLessOccurs(String[] list) {
        return getListOfLessOccurs(Arrays.asList(list), 0);
    }

    private List<String> getListOfMostOccurs(List<String> list, int position) {
        return getListOf(list, position, (countZeros, countOnes) -> countZeros > countOnes);
    }

    private List<String> getListOfLessOccurs(List<String> list, int position) {
        return getListOf(list, position, (countZeros, countOnes) -> countZeros <= countOnes);
    }

    private List<String> getListOf(List<String> list, int position, ListFilter listFilter) {
        List<String> numbersWithZeros = new ArrayList<String>();
        List<String> numbersWithOnes = new ArrayList<String>();

        for (int i = 0; i < list.size(); i++) {
            char number = list.get(i).charAt(position);
            if (number == '0') {
                numbersWithZeros.add(list.get(i));
            } else {
                numbersWithOnes.add(list.get(i));
            }
        }

        int countZeros = numbersWithZeros.size();
        int countOnes = numbersWithOnes.size();
        List<String> nextList = listFilter.filter(countZeros, countOnes) ? numbersWithZeros : numbersWithOnes;
        if (nextList.size() == 1) {
            return nextList;
        } else {
            return getListOf(nextList, position + 1, listFilter);
        }
    }

    public static void main(String[] args) {
        new Day3().run(3);
    }

    interface ListFilter {
        boolean filter(int countZeros, int countOnes);
    }
}